firebase-firestore/ktx/src/main/kotlin/com/google/firebase/firestore/ktx/Firestore.kt

```diff
@@ -162,3 +171,73 @@ class FirebaseFirestoreKtxRegistrar : ComponentRegistrar {
     override fun getComponents(): List<Component<*>> =
             listOf(LibraryVersionComponent.create(LIBRARY_NAME, BuildConfig.VERSION_NAME))
 }
+
+/**
+ * Starts listening to the document referenced by this `DocumentReference` with the given options and emits its values via a [Flow].
+ *
+ * - When the returned flow starts being collected, an [EventListener] will be attached.
+ * - When the flow completes, the listener will be removed.
+ *
+ * Backpressure: by default the returned flow is conflated. If the consumer isn't fast enough,
+ * it might miss some values but is always guaranteed to get the latest value. Use the [bufferCapacity] parameter
+ * to change that behaviour.
+ *
+ * @param metadataChanges controls metadata-only changes. Default: [MetadataChanges.EXCLUDE]
+ * @param bufferCapacity the buffer capacity as in [Flow.buffer] or null to not buffer at all
+ */
+fun DocumentReference.snapshots(
+    metadataChanges: MetadataChanges = MetadataChanges.EXCLUDE,
+    bufferCapacity: Int? = Channel.CONFLATED
+): Flow<DocumentSnapshot> {
+    val flow = callbackFlow {
+        val registration = addSnapshotListener(metadataChanges) { snapshot, exception ->
+            if (exception != null) {
+                cancel(message = "Error getting DocumentReference snapshot", cause = exception)
+            } else if (snapshot != null) {
+                trySend(snapshot)
```

Comment:
shouldn't it be trySendBlocking instead? here and below

Reply:
see my response re `buffer`, as for the comment you linked:

> @martinbonnin Why not just use trySend instead of risking blocking the thread, which is likely to be the main one?

imo the fact that Firestore delivers events on the UI thread is an unfortunate(but necessary) historic artifact as it optimized for devEx before coroutines were available and switching to the UI thread was non-trivial, and most devs want to render UI when data changes so UI thread was chosen. However Firestore provides overloads for addSnapshotListener that takes an executor. Now in the coroutine and flow world, devs have explicit control of what thread they are collecting on, so we can afford to deliver events to the flow on any thread we like.

So I think it's a safer choice to be as unopinionated as possible and use a safe executor + trySendBlocking, i.e. AsyncTask.THREAD_POOL_EXECUTOR or `com.google.firebase.firestore.util.Executors.BACKGROUND_EXECUTOR`(which is more limited).

This will make our code safe by default, i.e. we will not "drop" events with CONFLATED, and given direct control to developers. wdyt?

cc @wu-hui @ehsannas

Reply:
Sounds good to me

Reply:
I've pushed a change to use `BACKGROUND_EXECUTOR`. 

2 questions though:

1. Isn't this the equivalent of `buffer(UNLIMITED)` in practice? It's not using the coroutines `buffer()` but I think it will enqueue `Runnable`s in the executors. So all in all it's also doing a choice for the user.
2. Is there any executor that could avoid a thread switch (immediate maybe?) ? I'm not familiar with the Firestore internals but if the code is running in a background thread already, feels like there's no need to dispatch?

Reply:
1. Yeah, I guess you're right, it may be suboptimal in this case. so direct executor sounds more appealing.
2. That's a good question, on the one hand I like the idea of using direct executor but I am not sure where the execution happen in that case. grpc io pool maybe? what is the risk of starving those threads in that case? it may not actually be an issue since the developer have full control and can avoid congestion of that pool.

Would love to hear thoughts from the firestore folks on this.

Reply:
I chatted with Firestore folks offline, given that Firestore is single-threaded internally, they really don't want to run arbitrary user code on their internal thread, so BACKGROUND_EXECUTOR is preferred.

@martinbonnin thanks for accommodating all the changes

Reply:
Sounds good 👍